### PR TITLE
[Rebase & FF] 202405: Changes related to upstream unit test code coverage

### PR DESCRIPTION
**Pull request by cfernald**
## Description

Merge changes not in upstream for code coverage in unit tests.

cherry-pick from 649b6c069f 11837561aa 857b4ed881 991a64e (pulled from 43ae607) 785fe3db51

- [x] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

N/A

## Integration Instructions

N/A

### CODE REVIEW

**Comment by cfernald** [comment]
Note for @os-d, dropped the pipelines changes in favor of a wholesale copy later.

**Review by makubacki — APPROVED** [review]

File: BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py
Comment (makubacki) [inline BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py:1]: Is there value in keeping the three commits separate? Can this just be a single HostBasedUnitTestRunner update?

File: UnitTestFrameworkPkg/ReadMe.md [hunk 1]
```diff
@@ -7,7 +7,6 @@ the UEFI shell environment and host-based environments. It allows for unit test
 on the tests and leave error logging, result formatting, context persistence, and test running to the framework.
 The unit test framework works well for low level unit tests as well as system level tests and
 fits easily in automation frameworks.
-
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: This can be dropped

**Review by os-d — APPROVED** [review]

File: BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py
Comment (cfernald) [reply BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py:1]: I guess it depends on what we want to take upstream. @Javagedes , what do you think about your changes here?

Comment (cfernald) [reply BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py:1]: Will squash joey's commits

**Pull request merged**